src/container.cpp

```diff
@@ -416,27 +414,27 @@ ReturnValue Container::queryRemove(const Thing& thing, uint32_t count, uint32_t
 	return RETURNVALUE_NOERROR;
 }
 
-Cylinder* Container::queryDestination(int32_t& index, const Thing& thing, Item** destItem, uint32_t& flags)
+std::shared_ptr<Thing> Container::queryDestination(int32_t& index, std::shared_ptr<const Thing> thing,
+                                                   std::shared_ptr<Item>& destItem, uint32_t& flags)
 {
 	if (!unlocked) {
-		*destItem = nullptr;
-		return this;
+		destItem = nullptr;
+		return shared_from_this();
 	}
 
 	if (index == 254 /*move up*/) {
 		index = INDEX_WHEREEVER;
-		*destItem = nullptr;
+		destItem = nullptr;
 
-		Container* parentContainer = dynamic_cast<Container*>(getParent());
-		if (parentContainer) {
-			return parentContainer;
+		if (auto parent = getParent()) {
```

Comment:
Could this be `const auto`
```suggestion
		if (const auto parent = getParent()) {
```